lib/Corres_UL.thy

```diff
@@ -1515,4 +1515,113 @@ lemma corres_assert_opt_assume_l:
   \<Longrightarrow> corres_underlying sr nf nf' dc (P and K (X \<noteq> None)) Q (assert_opt X >>= f) g"
   by (force simp: corres_underlying_def assert_opt_def return_def bind_def fail_def)
 
+section \<open> Preservation of state_relation for the concrete side symbolic execution \<close>
+
+(** sr_inv **)
+
+definition sr_inv_ul :: "('a \<times> 'b) set \<Rightarrow> ('a \<Rightarrow> bool) \<Rightarrow> ('b \<Rightarrow> bool) \<Rightarrow> ('b \<Rightarrow> (unit \<times> 'b) set \<times> 'c) \<Rightarrow> bool" where
+  "sr_inv_ul sr P P' f \<equiv>
+       \<forall>s s' t'. (s, s') \<in> sr \<longrightarrow> P s \<longrightarrow> P' s'
+              \<longrightarrow> ((), t') \<in> fst (f s') \<longrightarrow> (s, t') \<in> sr"
+
+lemma corres_noop_assm: (* check equivalence *)
+  "(\<forall>s. P s \<longrightarrow> \<lbrace>P' and (\<lambda>s'. (s, s') \<in> sr)\<rbrace> m \<lbrace>\<lambda>_ s'. (s, s') \<in> sr\<rbrace>) \<longleftrightarrow> sr_inv_ul sr P P' m"
+  by (fastforce simp: valid_def sr_inv_ul_def)
+
+lemma sr_inv_ul_imp:
+  "\<lbrakk>sr_inv_ul sr Q Q' f; \<And>s. P s \<Longrightarrow> Q s; \<And>s. P' s \<Longrightarrow> Q' s\<rbrakk> \<Longrightarrow> sr_inv_ul sr P P' f "
+  by (clarsimp simp: sr_inv_ul_def)
+
+lemma sr_inv_stronger_imp:
```

Comment:
I’d say these are for convenience, just like various corres lemmas of slightly different versions, or equivalent rewriting, etc.

Reply:
So, we have `corres_guard_imp` and `stronger_corres_guard_imp` (which `sr_inv_stronger_imp` is based on), but we almost always use the first unless it is required to use the other one, I think. But the stronger one is used to prove the first. So, I can do that, surely, but I think we'd want to keep the simpler one as the main thing to use.

Reply:
So, I will prove the simpler one using the stronger one, as both of you suggest, but for the rest and also for the comment below for the `_cross` lemmas, I am not sure.

For the corres rules, we use `corres_guard_imp`, `corres_cross_over_guard`, `corres_cross_add_guard`, and the like, and all are more or less immediate consequences of `stronger_corres_guard_imp`. And, each of these is used on specific situations that it caters for, and we currently use `stronger_corres_guard_imp` itself very rarely. Do you think we should give a shortest name to `stronger_corres_guard_imp` and use it for everything?

I wouldn't think so...because it is rare enough that we actually need the stronger version to complete the proof, and also because the corres subgoals are generally quite complex, we can make use of simpler versions, when we can, which is quite often, to have simpler and clearer proof structure. And that is what we have been doing, I believe.

For `sr_inv`, in particular, it is often possible to extract information from the state relation and the abstract side at the corres level before getting to the symbolic execution. So, it is also likely that the simpler version suffices most of the time. In the proof of `reply_remove_tcb_corres`, there is one subcase that I squeezed all the case distinctions inside one symbolic execution step, and that is the only instance that I needed the stronger version `sr_inv_stronger_imp`.

Reply:
Having said that, I imagine `sr_inv` goals are generally smaller than the corres ones, and we won't have so many use cases of `sr_inv` anyway, so it probably doesn't' really make much difference either way.